src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs

```diff
@@ -0,0 +1,2107 @@
+                Diagnostic(ErrorCode.ERR_AssignReadonlyNotField, "F").WithArguments("field", "S<T>.F").WithLocation(23, 13),
+                // (24,13): error CS8331: Cannot assign to field 'S<T>.F' because it is a readonly variable
+                //             F = GetRef();
+                Diagnostic(ErrorCode.ERR_AssignReadonlyNotField, "F").WithArguments("field", "S<T>.F").WithLocation(24, 13),
+                // (25,13): error CS8331: Cannot assign to field 'S<T>.F' because it is a readonly variable
+                //             F = GetRefReadonly();
+                Diagnostic(ErrorCode.ERR_AssignReadonlyNotField, "F").WithArguments("field", "S<T>.F").WithLocation(25, 13),
+                // (41,9): error CS8331: Cannot assign to field 'S<T>.F' because it is a readonly variable
+                //         s.F =  tValue;
+                Diagnostic(ErrorCode.ERR_AssignReadonlyNotField, "s.F").WithArguments("field", "S<T>.F").WithLocation(41, 9),
+                // (42,9): error CS8331: Cannot assign to field 'S<T>.F' because it is a readonly variable
+                //         s.F =  tRef;
+                Diagnostic(ErrorCode.ERR_AssignReadonlyNotField, "s.F").WithArguments("field", "S<T>.F").WithLocation(42, 9),
+                // (43,9): error CS8331: Cannot assign to field 'S<T>.F' because it is a readonly variable
+                //         s.F =  tOut;
+                Diagnostic(ErrorCode.ERR_AssignReadonlyNotField, "s.F").WithArguments("field", "S<T>.F").WithLocation(43, 9),
+                // (44,9): error CS8331: Cannot assign to field 'S<T>.F' because it is a readonly variable
+                //         s.F =  tIn;
+                Diagnostic(ErrorCode.ERR_AssignReadonlyNotField, "s.F").WithArguments("field", "S<T>.F").WithLocation(44, 9));
+        }
+
+        [Fact]
+        public void Assignment_ReadonlyRef()
+        {
+            var source =
+@"ref struct S<T>
+{
+    public readonly ref T F;
+    public S(T tValue, ref T tRef, out T tOut, in T tIn)
+    {
+        tOut = default;
+        F = ref tValue;
+        F = ref tRef;
+        F = ref tOut;
+        F = ref tIn;
+        F = tValue;
+        F = tRef;
+        F = tOut;
+        F = tIn;
+    }
+    object P
+    {
+        init
+        {
+            F = ref GetValue();
+            F = ref GetRef();
+            F = ref GetRefReadonly();
+            F = GetValue();
+            F = GetRef();
+            F = GetRefReadonly();
+        }
+    }
+    static T GetValue() => throw null;
+    static ref T GetRef() => throw null;
+    static ref readonly T GetRefReadonly() => throw null;
+}
+class Program
+{
+    static void Assign<T>(S<T> s, T tValue, ref T tRef, out T tOut, in T tIn)
+    {
+        tOut = default;
+        s.F =  ref tValue;
+        s.F =  ref tRef;
+        s.F =  ref tOut;
+        s.F =  ref tIn;
+        s.F =  tValue;
+        s.F =  tRef;
+        s.F =  tOut;
+        s.F =  tIn;
```

Comment:
Whether these are legal or not depends on how we want to interpret `readonly ref`. Is it effectively a short hand for `readonly ref readonly` or not? I think it's more consistent with other features, think for example how `scope` is defined, for it to be simply `readonly ref`. Essentially a `ref` which cannot be re-pointed but where the underlying value can be mutated.